### PR TITLE
Integrating astropy.time into cdflib

### PR DESCRIPTION
**Pull request by jibarnum**
Response to ticket #14 

Integrated astropy.time into cdflib where possible, converted cdflib.epochs from class to set of functions (as the utility of a class was not used at all), cleaned up related tests and changed other scripts based on these changes.

### CODE REVIEW

**Comment by dstansby** [comment]
Hmm, looks like this still has conflicts. Is there any way it could be split up into several smaller pull requests too? That would make it much easier to review.

File: cdflib/epochs.py [hunk 1]
```diff
@@ -1290,523 +940,322 @@ def epochrange_epoch16(epochs, starttime=None, endtime=None):  # @NoSelf
...
             new_dates = dates
         count = len(new_dates)
-        if (count == 0):
+        if count == 0:
             return None
-        epochs = []
+        epochs = list()
         for x in range(0, count):
             date = new_dates[x]
-            year = date[0]
```
Comment (bryan-harter) [inline cdflib/epochs.py]: Looks like this function lost its ability to take in lists less than 7?

File: README.rst [hunk 2]
```diff
@@ -0,0 +1,635 @@
+This is a python script to read/write CDF files without needing to install the CDF NASA library.
```
Comment (bryan-harter) [inline README.rst]: I don't think we need this rst document

**Comment by bryan-harter** [comment]
A few things with this ticket, I'll make separate comments that might all become separate tickets.  


For the time being I would make a new astropy time class/file where the astropy code resides.  So in otherwords, just make like a CDFAstropyEpoch class/file that contains the class class you are writing.  This is in addition to the current CDFEpoch class on master.  

And when importing the epoch class, you can have some code that looks like:

```
try:
    import astropy
    import cdflib.astroepoch as epoch
except:
    import cdflib.epoch as epoch
```


Basically, that's because I think that absolutely requiring users to have astropy for cdflib to run might not be ideal.  I know of at least one user out there that is a geologists, and geologists would probably not ever have astropy installed.  Then again, astropy is a relatively easy installation, and we could just simply add it to the requirements file. 

But getting rid of the old method that only required numpy in just one new version might be too quick of a change.  Plus, having both options available in the same code makes it easy to test that both astropy and the old methods give the same answer.

**Comment by bryan-harter** [comment]
I think a good way to help with all the conversions would be to make astropy time format classes.  I tested out the following classes and they seem to work out pretty well:

(Used this website to help test: https://www.epochconverter.com/seconds-days-since-y0 )

```
from astropy.time import Time, TimeFormat
from astropy.time.formats import erfa, TimeFromEpoch


class TimeEpoch(TimeFromEpoch):
    name = 'cdf_epoch'
    unit = 1.0 / (erfa.DAYSEC * 1000) # Milliseconds
    epoch_val = '0000-01-01 00:00:00'
    epoch_val2 = None
    epoch_scale = 'utc'
    epoch_format = 'iso'

class TimeEpoch16(TimeFromEpoch):
    name = 'cdf_epoch16'
    unit = 1.0 / (erfa.DAYSEC) # Seconds
    epoch_val = '0000-01-01 00:00:00'
    epoch_val2 = None
    epoch_scale = 'utc'
    epoch_format = 'iso'

class TimeTT2000(TimeFromEpoch):
    name = 'cdf_tt2000'
    unit = 1.0 / (erfa.DAYSEC * 1000 * 1000 * 1000) #Nanoseconds
    epoch_val = '2000-01-01 12:00:00'
    epoch_val2 = None
    epoch_scale = 'tt'
    epoch_format = 'iso'
```

These could be used in the following manner:

```
#Convert from CDF_EPOCH to ISO String and Unix
t = Time(62285326000000.0, format='cdf_epoch', precision=6)
print(t.unix)
print(t.iso)

#Convert from CDF_EPOCH16 to ISO/UNIX
#Note: the first number is seconds since year 0, second number is the decimal precision
#Normally, the second number of a CDF_EPOCH16 is actually picoseconds.  Need to divide by 10^12 to turn into units of seconds
t = Time(33300946758.0, 0.106014648000, format='cdf_epoch16', precision=9)
print(t.iso)

#Convert from Unix time to CDF_TT2000
t = Time([946728435.816, 1070184724.917112], format='unix')
print(t.cdf_tt2000)
```

Using these methods, it should be pretty easy to turn any of the CDF times into unix time, ISO strings, or have them broken down into the list ['year', 'month', 'day', etc].  It also works the other way as well, you can read data into Astropy Time objects, and with these classes you can just print them out as CDF_EPOCH/CDF_EPOCH16/CDF_TT2000.

**Comment by bryan-harter** [comment]
I think we could use functions that convert between CDF times and Astropy times and vice versa.  Right now we have 

Breakdown() <---> Compute() for transforming between lists of numbers and CDF times

parse() <---> encode() for transforming between ISO strings and CDF times

unixtime() for converting CDF times to seconds since 1970 (should probably also have a function to convert unixtime into cdf times)

We should add functions for back and forth conversions between Astropy times and CDF times too.

**Comment by bryan-harter** [comment]
The only problem I can foresee with switching entirely to using astropy in this library for conversions is that CDF_EPOCH16 has a precision of picoseconds, and I don't believe astropy can guarantee that.  There may be a way around this, but the highest precision I can get astropy to print out is 9 digits after the decimal, so nanosecond precision.   

And, using the above code, a conversion like this:

```
t = Time(946728435.816, format='unix', precision=9)
print(t.cdf_epoch16)
```

prints out only 1 number, so you lose a LOT of precision there.  I wonder if there is a way to force it to print out 2 numbers, since I know it uses two 64 bit numbers to store the time.  Perhaps buried in the documentation lies the answer.

**Comment by scivision** [comment]
There are some underlying structural things broken/antipatterns with cdflib that should be resolved incrementally that we don't want to keep preserving.

* constant redownloading of leapsecond #47 
* need to raise Exceptions e.g. TypeError and KeyError instead of `return None`
* need to fix code style to PEP8
* need to fix default function arguments
* need to clean up syntax to Python style instead of C style

I have PRs open for these, and more fixes once those PRs are accepted.

I also think we shouldn't require anything except Numpy.

However we need to make a lot of simple fixes, that I've already done a year or so ago, that need to be merged here to help avoid further cdflib fragmentation between here and my fork of "fixed" cdflib.

Once we do that, and the work is already done, just needs merging, it will be easier to add optional features like Astropy.time

**Comment by bryan-harter** [comment]
I made an optional module for people to use instead of the primary epochs module.  There are still a few issues though, and I made new tickets for those issues.

**Pull request closed without merging**